src/Illuminate/Database/Grammar.php

```diff
@@ -251,4 +255,26 @@ public function setTablePrefix($prefix)
 
         return $this;
     }
+
+    /**
+     * Quote the given string literal.
+     *
+     * @param  string  $value
+     * @return string
+     */
+    abstract protected function quoteValue($value);
```

Comment:
Do these 3 methods need to be abstract methods? They are never actually implemented except by the traits from what I can tell? Why can the base grammar (this file) not simply use the traits? Why does each grammar import the traits individually?

Reply:
The implementation is built with extensibility in mind. If the base grammar has to implement `quoteValue` it needs to be tied to a specific implementation. In this case, any grammar will have a hard dependency to `PDO`, which is used by every Laravel database driver at the moment. But 3rd party drivers like [jenssegers/laravel-mongodb](https://github.com/jenssegers/laravel-mongodb) are not using `PDO`. Moving the logic of `quoteValue ` from the trait to the base `Grammar` class prevents anyone from implementing 3rd party db drivers which are not built on top of `PDO`. So in this change I did the effort to ensure backward compatibility for 3rd party packages.